### PR TITLE
Use task to offload work to background thread

### PR DESCRIPTION
**Pull request by gingters**
Merge #349 first

### CODE REVIEW

File: src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs [hunk 1]
```diff
@@ -48,66 +45,36 @@ public class ClientRequestHandler<TRequest, TResponse, TAcknowledge> : IClientRe
...
 
-			if (request.AcknowledgeMode == AcknowledgeMode.Manual)
-			{
-				var url = new Uri(_acknowledgeEndpoint, $"{request.AcknowledgeOriginId}/{request.RequestId}").ToString();
-				request.HttpHeaders[Constants.HeaderNames.AcknowledgeUrl] = new[] { url };
-			}
+				if (request.AcknowledgeMode == AcknowledgeMode.Manual)
+				{
+					var url = new Uri(_acknowledgeEndpoint, $"{request.AcknowledgeOriginId}/{request.RequestId}").ToString();
+					request.HttpHeaders[Constants.HeaderNames.AcknowledgeUrl] = new[] { url, };
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs:69]:
```suggestion
					request.HttpHeaders[Constants.HeaderNames.AcknowledgeUrl] = new[] { url };
```

File: src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs [hunk 2]
```diff
@@ -48,66 +45,36 @@ public class ClientRequestHandler<TRequest, TResponse, TAcknowledge> : IClientRe
...
 
-			if (request.EnableTracing)
-			{
-				request.HttpHeaders[Constants.HeaderNames.RequestId] = new[] { request.RequestId.ToString() };
-				request.HttpHeaders[Constants.HeaderNames.OriginId] = new[] { request.RequestOriginId.ToString() };
-			}
+				if (request.EnableTracing)
+				{
+					request.HttpHeaders[Constants.HeaderNames.RequestId] = new[] { request.RequestId.ToString(), };
+					request.HttpHeaders[Constants.HeaderNames.OriginId] = new[] { request.RequestOriginId.ToString(), };
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs:75]:
```suggestion
					request.HttpHeaders[Constants.HeaderNames.RequestId] = new[] { request.RequestId.ToString() };
					request.HttpHeaders[Constants.HeaderNames.OriginId] = new[] { request.RequestOriginId.ToString() };
```

File: src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs [hunk 3]
```diff
@@ -128,8 +95,8 @@ private async Task DeliverResponseAsync(TResponse response, bool enableTracing)
 			if (enableTracing)
 			{
 				response.HttpHeaders ??= new Dictionary<string, string[]>();
-				response.HttpHeaders[Constants.HeaderNames.ConnectorMachineName] = new[] { Environment.MachineName };
-				response.HttpHeaders[Constants.HeaderNames.ConnectorVersion] = new[] { RelayConnector.AssemblyVersion };
+				response.HttpHeaders[Constants.HeaderNames.ConnectorMachineName] = new[] { Environment.MachineName, };
+				response.HttpHeaders[Constants.HeaderNames.ConnectorVersion] = new[] { RelayConnector.AssemblyVersion, };
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs:99]:
```suggestion
				response.HttpHeaders[Constants.HeaderNames.ConnectorMachineName] = new[] { Environment.MachineName };
				response.HttpHeaders[Constants.HeaderNames.ConnectorVersion] = new[] { RelayConnector.AssemblyVersion };
```

**Review by thomashilzendegen — CHANGES_REQUESTED** [review]
There are some commas which should not be there (only on multiline).

**Comment by gingters** [comment]
Changes done and squashed in the original commit

File: src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs [hunk 4]
```diff
@@ -48,66 +45,36 @@ public class ClientRequestHandler<TRequest, TResponse, TAcknowledge> : IClientRe
...
-		{
-			if (QueueWorker(request, cancellationToken)) return;
-
-			if (request.AcknowledgeMode == AcknowledgeMode.ConnectorReceived)
-			{
-				await AcknowledgeRequestAsync(request, false);
-			}
-
-			await DeliverResponseAsync(request.CreateResponse<TResponse>(HttpStatusCode.ServiceUnavailable), request.EnableTracing);
+			Task.Run(() => WorkerCallAsync(request, cancellationToken), cancellationToken);
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs:53]:
```suggestion
 			if (request.AcknowledgeMode == AcknowledgeMode.ConnectorReceived)
 			{
 				await AcknowledgeRequestAsync(request, false);
 			}

			Task.Run(() => WorkerCallAsync(request, cancellationToken), cancellationToken);
```

**Review by thomashilzendegen — CHANGES_REQUESTED** [review]

File: src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs [hunk 5]
```diff
@@ -48,66 +45,36 @@ public class ClientRequestHandler<TRequest, TResponse, TAcknowledge> : IClientRe
...
-				ThreadPool.SetMaxThreads(workerThreads, _maxCompletionPortThreads);
-			}
-		}
-
-		/// <inheritdoc />
-		public async Task HandleAsync(TRequest request, CancellationToken cancellationToken = default)
-		{
-			if (QueueWorker(request, cancellationToken)) return;
-
-			if (request.AcknowledgeMode == AcknowledgeMode.ConnectorReceived)
```
[the remark this answers is not in the document]
Comment (gingters) [reply src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs:79]: That already happens in the Worker, so this was a duplicate ack: https://github.com/thinktecture/relayserver/blob/feature/background-threads/src/Thinktecture.Relay.Connector/Targets/ClientRequestWorker.cs#L138-L141

File: src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs (hunk 4 above)
Comment (gingters) [reply src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs:53]: That would duplicate the ConnectorReceived ACK again ;)

Comment (gingters) [reply src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs:53]: We would need to decide if it should be done by the handler before handing the request to the worker, or by the worker in the finally block. The difference would be that atm the ack might be delayed by the worker's timeout when requesting the target.

Comment (thomashilzendegen) [reply src/Thinktecture.Relay.Connector/Targets/ClientRequestHandler.cs:53]: Finally it is!

**Review by thomashilzendegen — APPROVED** [review]

**Pull request merged**